### PR TITLE
Side Effects and Mutable Side Effects

### PR DESCRIPTION
**Pull request by rachfop**
### Side Effect
Adds Workflow Side Effect for:
- [ ] Go
- [ ] Java

Python and TypeScript do not currently support Side Effects.

https://github.com/temporalio/features/issues/561

### Mutable Side Effect
Adds Mutable Side Effect for:
- [ ] Go
- [ ] Java
- [ ] PHP
- [ ] Python
- [ ] TypeScript

## NOTE

I didn't touch any of the random files in this PR.

It's probably from running `yarn gen` and Update branch.

### CODE REVIEW

File: docs/app-dev-context/side-effects.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+---
+id: side-effects
+title: Side Effects
```
Comment (djmagee) [inline docs/app-dev-context/side-effects.md:3]: Should we combine this file with `\concepts\what-is-a-side-effect` and then use the one file both here and in the Glossary?

File: docs/app-dev-context/side-effects.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+description: Side Effects are a method of execution to produce nondeterministic code.
+sidebar_label: Side Effects
+tags:
+  - guide-context
+---
+
+A Side Effect is a method of execution to produce nondeterministic code; such as, generating a UUID or a random number.
+By implementing Side Effect into your Workflow Execution, you can execute the provided function once and records its result into the Workflow Execution [Event History](../concepts/what-is-an-event-history).
```
Comment (djmagee) [inline docs/app-dev-context/side-effects.md:11]:
```suggestion
A Side Effect is a method of execution to produce nondeterministic code, such as generating a UUID or a random number.
By implementing a Side Effect in your Workflow Execution, you can execute the provided function once and record its result into the Workflow Execution [Event History](/workflows/#event-history).
```

File: docs/go/how-to-use-a-side-effect-in-go.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+var random int
+encodedRandom.Get(&random)
+if random < 50 {
+       ....
+} else {
+       ....
+}
+```
+
+The only way to fail Side Effect is to panic, which causes Workflow Task failure. The Workflow Task after Timeout is rescheduled and re-executed giving Side Effect another chance to succeed. Be careful to not return any data from the Side Effect function any other way than through its recorded return value.
```
Comment (djmagee) [inline docs/go/how-to-use-a-side-effect-in-go.md:26]:
```suggestion
The only way to fail a Side Effect is to panic, which causes Workflow Task failure.
The Workflow Task after Timeout is rescheduled and re-executed giving the Side Effect another chance to succeed.
Be careful not to return any data from the Side Effect function any way other than through its recorded return value.
```

File: docs/go/how-to-use-a-side-effect-in-go.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+Do not use `SideEffect()` to modify closures. Always retrieve result from Side Effect's encoded return value.
```
Comment (djmagee) [inline docs/go/how-to-use-a-side-effect-in-go.md:28]:
```suggestion
Do not use `SideEffect()` to modify closures.
Always retrieve results from a Side Effect's encoded return value.
```

File: docs/java/how-to-use-a-side-effect-in-java.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+```java
+  int random = Workflow.sideEffect(Integer.class, () -> random.nextInt(100));
+  if random < 50 {
+         ....
+  } else {
+         ....
+  }
+```
+
+Another example using `sideEffect()`.
```
Comment (djmagee) [inline docs/java/how-to-use-a-side-effect-in-java.md:22]:
```suggestion
Here's another example that uses `sideEffect()`.
```

**Comment by cretz** [comment]
How sure are we that we want side effects and mutable side effects documented in the app dev guide? They are Java/Go/PHP only and other languages may never implement them. There are lots of features of specific APIs we don't document in the app dev guide and instead refer to the SDK documentation.

Can we get business/product confirmation that this is something we want to document in our cross-language guide? If we do, can business/product then provide the text making it clear this is not available in TypeScript/Python by intention?

**Comment by cretz** [comment]
Not sure we should document this in the cross-language app dev guide until we're sure we're ever even going to have this feature across languages. If we choose not to have this feature across languages, we need to decide whether this is worth documenting at all and/or what the language would be saying why other languages don't have it.

**Comment by Spikhalskiy** [comment]
It does definitely need to be documented as a side effect is the only safe/correct way of working with external configuration in Workflow code. Using local activities for it is a temporary hack. We should discuss though if we exposed a safer / more user-friendly interface over it, like a "Marker", because side-effect is a little technical and, actually, misleading.

**Comment by cretz** [comment]
@Spikhalskiy - TS impl issue at https://github.com/temporalio/sdk-typescript/issues/144 is where we could potentially discuss different name or approach (no issue for Core or Python yet, it'd come from that). Then we can come back and update the doc stuff. Unsure if we need to leave this doc PR on hold until that is discussed.

**Comment by flossypurse** [comment]
We have already decided that it is perfectly acceptable to document a feature that is not at parity with the other SDKs...
We simply indicate that the feature is not currently available with that SDK.

File: docs/go/how-to-use-a-side-effect-in-go.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
+---
+id: how-to-use-a-side-effect-in-go
```
Comment (flossypurse) [inline docs/go/how-to-use-a-side-effect-in-go.md:2]: Were you aware of the existing file for this?
https://github.com/temporalio/documentation/blob/master/docs/go/how-to-execute-a-side-effect-in-go.md

tbh - I think the existing one is more to the point and clearer atm

**Comment by cretz** [comment]
@flossypurse - :+1: My concern was whether this feature is important enough to document in the guide. Note, the app guide doesn't document a dozen Go SDK features (await, selectors, channels, etc) and every SDK has many features we choose not to document there. If we feel side effects aren't deserving of this kind of documentation and may just add more noise than value, we can leave it up to the API documentation like we do with _lots_ of other features.

**Comment by flossypurse** [comment]
@cretz 
> Note, the app guide doesn't document a dozen Go SDK features (await, selectors, channels, etc)

Valid point are these features covered by the sdk-features repo?

Are there other "concurrency" type of features in other SDKs that are somewhat on parallel?
For example, is threading possible in Java, and do we expose a specific wrapper API to do that in Java like we do with Go channels?

@rachfop can you remind me if this is being tracked anywhere atm or if we made a decision about this yet?

**Comment by cretz** [comment]
> Valid point - are these features covered by the sdk-features repo? [...] Are there other "concurrency" type of features in other SDKs that are somewhat on parallel?

Not all of them, or not yet. But for example, all SDKs provide a "wait for condition" type of API that we don't document. Same with timers/sleep (or at least I can't find the docs). Same for starting a coroutine.

Basically the actual API documents lots of things the docs doesn't (and probably shouldn't). Some are features all SDKs have, some are features only some SDKs have, some are features only one SDK has. The decision to document should just be based on importance vs noise.

> For example, is threading possible in Java, and do we expose a specific wrapper API to do that in Java like we do with Go channels?

For channels, Java does have `Workflow.newQueue` which provides similar behavior to Go's `workflow.NewBufferedChannel`. For threading/coroutines, Java does have `Async.function`/`Async.procedure` which provides similar behavior to Go's `workflow.Go`.

I don't believe we document any of this (and again, maybe we shouldn't, but maybe we should, unsure).

**Pull request closed without merging**